File: Shoply/Views/DesignSystem.swift
//
//  DesignSystem.swift
//  Shoply - Outfit Selector
//
//  Created by William on 01/11/2025.
//

import SwiftUI

// MARK: - Système de Design Épuré Noir & Blanc

struct DesignSystem {
    
    // MARK: - Espacements
    struct Spacing {
        static let xs: CGFloat = 4
        static let sm: CGFloat = 8
        static let md: CGFloat = 16
        static let lg: CGFloat = 24
        static let xl: CGFloat = 32
        static let xxl: CGFloat = 40
    }
    
    // MARK: - Rayons de coins
    struct Radius {
        static let sm: CGFloat = 8
        static let md: CGFloat = 12
        static let lg: CGFloat = 16
        static let xl: CGFloat = 20
    }
    
    // MARK: - Typographie
    struct Typography {
        static func largeTitle() -> Font {
            .system(size: 34, weight: .bold, design: .default)
        }
        
        static func title() -> Font {
            .system(size: 28, weight: .bold, design: .default)
        }
        
        static func title2() -> Font {
            .system(size: 22, weight: .semibold, design: .default)
        }
        
        static func headline() -> Font {
            .system(size: 17, weight: .semibold, design: .default)
        }
        
        static func body() -> Font {
            .system(size: 17, weight: .regular, design: .default)
        }
        
        static func callout() -> Font {
            .system(size: 16, weight: .regular, design: .default)
        }
        
        static func subheadline() -> Font {
            .system(size: 15, weight: .regular, design: .default)
        }
        
        static func footnote() -> Font {
            .system(size: 13, weight: .regular, design: .default)
        }
        
        static func caption() -> Font {
            .system(size: 12, weight: .regular, design: .default)
        }
    }
}

// MARK: - Composants Réutilisables

struct Card<Content: View>: View {
    @Environment(\.colorScheme) var colorScheme
    let content: Content
    let cornerRadius: CGFloat
    
    init(cornerRadius: CGFloat = DesignSystem.Radius.md, @ViewBuilder content: () -> Content) {
        self.cornerRadius = cornerRadius
        self.content = content()
    }
    
    var body: some View {
        content
            .padding(DesignSystem.Spacing.md)
            .background {
                RoundedRectangle(cornerRadius: cornerRadius)
                    .fill(.ultraThinMaterial)
            }
            .clipShape(RoundedRectangle(cornerRadius: cornerRadius))
    }
}

struct PrimaryButton: View {
    let title: String
    let action: () -> Void
    
    var body: some View {
        Button(action: action) {
            Text(title)
                .font(DesignSystem.Typography.headline())
                .foregroundColor(AppColors.buttonPrimaryText)
                .frame(maxWidth: .infinity)
                .padding(.vertical, DesignSystem.Spacing.md)
                .background(AppColors.buttonPrimary)
                .clipShape(RoundedRectangle(cornerRadius: DesignSystem.Radius.md))
        }
    }
}

struct SecondaryButton: View {
    let title: String
    let action: () -> Void
    
    var body: some View {
        Button(action: action) {
            Text(title)
                .font(DesignSystem.Typography.headline())
                .foregroundColor(AppColors.buttonSecondaryText)
                .frame(maxWidth: .infinity)
                .padding(.vertical, DesignSystem.Spacing.md)
                .background(AppColors.buttonSecondary)
                .clipShape(RoundedRectangle(cornerRadius: DesignSystem.Radius.md))
        }
    }
}

struct SectionHeader: View {
    let title: String
    
    var body: some View {
        HStack {
            Text(title)
                .font(DesignSystem.Typography.title2())
                .foregroundColor(AppColors.primaryText)
            Spacer()
        }
        .padding(.horizontal, DesignSystem.Spacing.md)
        .padding(.vertical, DesignSystem.Spacing.sm)
    }
}

struct ListRow<Content: View>: View {
    let content: Content
    
    init(@ViewBuilder content: () -> Content) {
        self.content = content()
    }
    
    var body: some View {
        HStack(spacing: DesignSystem.Spacing.md) {
            content
        }
        .padding(.horizontal, DesignSystem.Spacing.md)
        .padding(.vertical, DesignSystem.Spacing.md)
        .background(AppColors.cardBackground)
    }
}

struct EmptyState: View {
    let icon: String
    let title: String
    let message: String
    
    var body: some View {
        VStack(spacing: DesignSystem.Spacing.md) {
            Image(systemName: icon)
                .font(.system(size: 48, weight: .light))
                .foregroundColor(AppColors.secondaryText)
            
            Text(title)
                .font(DesignSystem.Typography.title2())
                .foregroundColor(AppColors.primaryText)
            
            Text(message)
                .font(DesignSystem.Typography.body())
                .foregroundColor(AppColors.secondaryText)
                .multilineTextAlignment(.center)
                .padding(.horizontal, DesignSystem.Spacing.xl)
        }
        .padding(.vertical, DesignSystem.Spacing.xxl)
    }
}

File: ShoplyRN/src/theme/DesignSystem.ts
import { Platform, StyleSheet } from 'react-native';

// MARK: - Espacements
export const Spacing = {
  xs: 4,
  sm: 8,
  md: 16,
  lg: 24,
  xl: 32,
  xxl: 40,
};

// MARK: - Rayons de coins
export const Radius = {
  sm: 8,
  md: 12,
  lg: 16,
  xl: 20,
};

// MARK: - Typographie
export const Typography = {
  largeTitle: {
    fontSize: 34,
    fontWeight: '700' as const,
    lineHeight: 41,
  },
  title: {
    fontSize: 28,
    fontWeight: '700' as const,
    lineHeight: 34,
  },
  title2: {
    fontSize: 22,
    fontWeight: '600' as const,
    lineHeight: 28,
  },
  headline: {
    fontSize: 17,
    fontWeight: '600' as const,
    lineHeight: 22,
  },
  body: {
    fontSize: 17,
    fontWeight: '400' as const,
    lineHeight: 22,
  },
  callout: {
    fontSize: 16,
    fontWeight: '400' as const,
    lineHeight: 21,
  },
  subheadline: {
    fontSize: 15,
    fontWeight: '400' as const,
    lineHeight: 20,
  },
  footnote: {
    fontSize: 13,
    fontWeight: '400' as const,
    lineHeight: 18,
  },
  caption: {
    fontSize: 12,
    fontWeight: '400' as const,
    lineHeight: 16,
  },
};

// MARK: - Couleurs (Noir & Blanc pour iOS, One UI 7 pour Android)
export const AppColors = {
  // Fond
  background: Platform.select({
    ios: '#FFFFFF',
    android: '#FFFFFF', // One UI 7 utilise aussi le blanc en mode clair
  }),
  backgroundDark: Platform.select({
    ios: '#000000',
    android: '#1C1C1E', // One UI 7 dark mode
  }),

  // Textes
  primaryText: Platform.select({
    ios: '#000000',
    android: '#000000',
  }),
  primaryTextDark: Platform.select({
    ios: '#FFFFFF',
    android: '#FFFFFF',
  }),
  secondaryText: Platform.select({
    ios: '#4D4D4D',
    android: '#6E6E73', // One UI 7 secondary text
  }),
  secondaryTextDark: Platform.select({
    ios: '#B3B3B3',
    android: '#AEAEB2',
  }),
  tertiaryText: Platform.select({
    ios: '#808080',
    android: '#8E8E93',
  }),

  // Cartes
  cardBackground: Platform.select({
    ios: '#FFFFFF',
    android: '#FFFFFF',
  }),
  cardBackgroundDark: Platform.select({
    ios: '#0D0D0D',
    android: '#2C2C2E', // One UI 7 card background
  }),
  cardBorder: Platform.select({
    ios: '#333333',
    android: '#E5E5EA',
  }),
  cardBorderDark: Platform.select({
    ios: '#333333',
    android: '#38383A',
  }),
  separator: Platform.select({
    ios: '#1A1A1A',
    android: '#C6C6C8',
  }),
  separatorDark: Platform.select({
    ios: '#1A1A1A',
    android: '#38383A',
  }),

  // Boutons
  buttonPrimary: Platform.select({
    ios: '#000000',
    android: '#000000',
  }),
  buttonPrimaryDark: Platform.select({
    ios: '#FFFFFF',
    android: '#FFFFFF',
  }),
  buttonPrimaryText: Platform.select({
    ios: '#FFFFFF',
    android: '#FFFFFF',
  }),
  buttonPrimaryTextDark: Platform.select({
    ios: '#000000',
    android: '#000000',
  }),
  buttonSecondary: Platform.select({
    ios: '#F2F2F7',
    android: '#F2F2F7',
  }),
  buttonSecondaryDark: Platform.select({
    ios: '#292929',
    android: '#2C2C2E', // One UI 7 secondary button
  }),
  buttonSecondaryText: Platform.select({
    ios: '#000000',
    android: '#000000',
  }),
  buttonSecondaryTextDark: Platform.select({
    ios: '#FFFFFF',
    android: '#FFFFFF',
  }),

  // Accents
  accent: Platform.select({
    ios: '#333333',
    android: '#333333',
  }),
  accentDark: Platform.select({
    ios: '#CCCCCC',
    android: '#CCCCCC',
  }),
  shadow: Platform.select({
    ios: 'rgba(0, 0, 0, 0.1)',
    android: 'rgba(0, 0, 0, 0.1)',
  }),
  shadowDark: Platform.select({
    ios: 'rgba(0, 0, 0, 0.3)',
    android: 'rgba(0, 0, 0, 0.3)',
  }),
};

// MARK: - Styles de base
export const baseStyles = StyleSheet.create({
  container: {
    flex: 1,
  },
  card: {
    borderRadius: Radius.md,
    padding: Spacing.md,
    backgroundColor: AppColors.cardBackground,
  },
  cardDark: {
    backgroundColor: AppColors.cardBackgroundDark,
  },
  primaryButton: {
    backgroundColor: AppColors.buttonPrimary,
    borderRadius: Radius.md,
    paddingVertical: Spacing.md,
    paddingHorizontal: Spacing.lg,
    alignItems: 'center',
    justifyContent: 'center',
  },
  primaryButtonDark: {
    backgroundColor: AppColors.buttonPrimaryDark,
  },
  primaryButtonText: {
    color: AppColors.buttonPrimaryText,
    ...Typography.headline,
  },
  primaryButtonTextDark: {
    color: AppColors.buttonPrimaryTextDark,
  },
  secondaryButton: {
    backgroundColor: AppColors.buttonSecondary,
    borderRadius: Radius.md,
    paddingVertical: Spacing.md,
    paddingHorizontal: Spacing.lg,
    alignItems: 'center',
    justifyContent: 'center',
  },
  secondaryButtonDark: {
    backgroundColor: AppColors.buttonSecondaryDark,
  },
  secondaryButtonText: {
    color: AppColors.buttonSecondaryText,
    ...Typography.headline,
  },
  secondaryButtonTextDark: {
    color: AppColors.buttonSecondaryTextDark,
  },
});

// MARK: - Helpers pour obtenir les couleurs selon le thème
export const getColors = (isDark: boolean) => ({
  background: isDark ? AppColors.backgroundDark : AppColors.background,
  primaryText: isDark ? AppColors.primaryTextDark : AppColors.primaryText,
  secondaryText: isDark ? AppColors.secondaryTextDark : AppColors.secondaryText,
  cardBackground: isDark ? AppColors.cardBackgroundDark : AppColors.cardBackground,
  cardBorder: isDark ? AppColors.cardBorderDark : AppColors.cardBorder,
  separator: isDark ? AppColors.separatorDark : AppColors.separator,
  buttonPrimary: isDark ? AppColors.buttonPrimaryDark : AppColors.buttonPrimary,
  buttonPrimaryText: isDark
    ? AppColors.buttonPrimaryTextDark
    : AppColors.buttonPrimaryText,
  buttonSecondary: isDark
    ? AppColors.buttonSecondaryDark
    : AppColors.buttonSecondary,
  buttonSecondaryText: isDark
    ? AppColors.buttonSecondaryTextDark
    : AppColors.buttonSecondaryText,
  shadow: isDark ? AppColors.shadowDark : AppColors.shadow,
});

